src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,341 @@
+        return DecodeIntoEmberBuffer<OddSizedInteger<7, true>>(decoder, isNullable, out);
+    case ZCL_INT64S_ATTRIBUTE_TYPE: // Signed 64-bit integer
+        return DecodeIntoEmberBuffer<int64_t>(decoder, isNullable, out);
+    case ZCL_SINGLE_ATTRIBUTE_TYPE: // 32-bit float
+        return DecodeIntoEmberBuffer<float>(decoder, isNullable, out);
+    case ZCL_DOUBLE_ATTRIBUTE_TYPE: // 64-bit float
+        return DecodeIntoEmberBuffer<double>(decoder, isNullable, out);
+    case ZCL_CHAR_STRING_ATTRIBUTE_TYPE: // Char string
+        return DecodeStringLikeIntoEmberBuffer<CharSpan, ShortPascalString>(decoder, isNullable, out);
+    case ZCL_LONG_CHAR_STRING_ATTRIBUTE_TYPE:
+        return DecodeStringLikeIntoEmberBuffer<CharSpan, LongPascalString>(decoder, isNullable, out);
+    case ZCL_OCTET_STRING_ATTRIBUTE_TYPE: // Octet string
+        return DecodeStringLikeIntoEmberBuffer<ByteSpan, ShortPascalString>(decoder, isNullable, out);
+    case ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE:
+        return DecodeStringLikeIntoEmberBuffer<ByteSpan, LongPascalString>(decoder, isNullable, out);
+    default:
+        ChipLogError(DataManagement, "Attribute type 0x%x not handled", static_cast<int>(metadata->attributeType));
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+    }
+    return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+}
+
+} // namespace
+
+CHIP_ERROR CodegenDataModel::WriteAttribute(const InteractionModel::WriteAttributeRequest & request,
+                                            AttributeValueDecoder & decoder)
+{
+    ChipLogDetail(DataManagement,
+                  "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=%x AttributeId=" ChipLogFormatMEI " (expanded=%d)",
+                  ChipLogValueMEI(request.path.mClusterId), request.path.mEndpointId, ChipLogValueMEI(request.path.mAttributeId),
+                  request.path.mExpanded);
+
+    // ACL check for non-internal requests
+    if (!request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        ReturnErrorCodeIf(!request.subjectDescriptor.has_value(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        Access::RequestPath requestPath{ .cluster = request.path.mClusterId, .endpoint = request.path.mEndpointId };
+        ReturnErrorOnFailure(Access::GetAccessControl().Check(*request.subjectDescriptor, requestPath,
+                                                              RequiredPrivilege::ForWriteAttribute(request.path)));
+    }
+
+    auto metadata = Ember::FindAttributeMetadata(request.path);
+
+    // Explicit failure in finding a suitable metadata
+    if (const CHIP_ERROR * err = std::get_if<CHIP_ERROR>(&metadata))
+    {
+        VerifyOrDie(*err != CHIP_NO_ERROR);
+        return *err;
+    }
+
+    // All the global attributes that we do not have metadata for are
+    // read-only (i.e. cannot write atribute_list/event_list/accepted_cmds/generated_cmds)
+    //
+    // so if no metadata available, we wil lreturn an error
```

Comment:
```suggestion
    // All the global attributes that we do not have metadata for are
    // read-only (i.e. cannot write attribute_list/accepted_cmds/generated_cmds, etc),
    // so if no metadata is available, we will return an error.
```

Note that this is a strong assumption. When we want to support code-backed clusters that report which attributes they support, lack of metadata may arise from EmberMetadata. Suggest adding a TODO comment with new issue  # for getting a non-ember-metadata source of truth for the presence of an attribute.

Reply:
I am working here specifically only for ember (i.e. codeggen) support, so there will never be a code-backed support here.

The implementation of code-backed bits will have to be separate, likely after global attribute handling (and ACL) is done inside IM based on metadata rather than the code here.